src/algebra/associated.lean

```diff
@@ -286,64 +294,66 @@ multiset.induction_on s (by simp [mt is_unit_iff_dvd_one.2 hp.not_unit])
-lemma ne_zero_iff_of_associated [monoid_with_zero α] {a b : α} (h : a ~ᵤ b) : a ≠ 0 ↔ b ≠ 0 :=
-not_congr $ eq_zero_iff_of_associated h
+lemma associated.ne_zero_iff [monoid_with_zero α] {a b : α} (h : a ~ᵤ b) : a ≠ 0 ↔ b ≠ 0 :=
+not_congr h.eq_zero_iff
 
-lemma prime_of_associated [comm_monoid_with_zero α] {p q : α} (h : p ~ᵤ q) (hp : prime p) :
+protected lemma associated.prime [comm_monoid_with_zero α] {p q : α} (h : p ~ᵤ q) (hp : prime p) :
   prime q :=
-⟨(ne_zero_iff_of_associated h).1 hp.ne_zero,
+⟨h.ne_zero_iff.1 hp.ne_zero,
   let ⟨u, hu⟩ := h in
     ⟨λ ⟨v, hv⟩, hp.not_unit ⟨v * u⁻¹, by simp [hv, hu.symm]⟩,
       hu ▸ by { simp [units.mul_right_dvd], intros a b, exact hp.div_or_div }⟩⟩
 
 lemma associated_of_irreducible_of_dvd [cancel_monoid_with_zero α] {p q : α}
   (p_irr : irreducible p) (q_irr : irreducible q) (dvd : p ∣ q) : associated p q :=
-associated_of_dvd_dvd dvd (dvd_symm_of_irreducible p_irr q_irr dvd)
+associated_of_dvd_dvd dvd (p_irr.dvd_symm q_irr dvd)
 
 lemma associated_of_prime_of_dvd [comm_cancel_monoid_with_zero α] {p q : α}
   (p_prime : prime p) (q_prime : prime q) (dvd : p ∣ q) : associated p q :=
-associated_of_irreducible_of_dvd (irreducible_of_prime p_prime) (irreducible_of_prime q_prime) dvd
+associated_of_irreducible_of_dvd p_prime.irreducible q_prime.irreducible dvd
 
-lemma prime_iff_of_associated [comm_monoid_with_zero α] {p q : α}
+lemma associated.prime_iff [comm_monoid_with_zero α] {p q : α}
   (h : p ~ᵤ q) : prime p ↔ prime q :=
-⟨prime_of_associated h, prime_of_associated h.symm⟩
+⟨h.prime, h.symm.prime⟩
+
+protected lemma associated.is_unit [monoid α] {a b : α} (h :  a ~ᵤ b) : is_unit a → is_unit b :=
+let ⟨u, hu⟩ := h in λ ⟨v, hv⟩, ⟨v * u, by simp [hv, hu.symm]⟩
 
-lemma is_unit_iff_of_associated [monoid α] {a b : α} (h :  a ~ᵤ b) : is_unit a ↔ is_unit b :=
-⟨let ⟨u, hu⟩ := h in λ ⟨v, hv⟩, ⟨v * u, by simp [hv, hu.symm]⟩,
-  let ⟨u, hu⟩ := h.symm in λ ⟨v, hv⟩, ⟨v * u, by simp [hv, hu.symm]⟩⟩
+lemma associated.is_unit_iff [monoid α] {a b : α} (h :  a ~ᵤ b) : is_unit a ↔ is_unit b :=
+⟨h.is_unit, h.symm.is_unit⟩
 
-lemma irreducible_of_associated [monoid α] {p q : α} (h : p ~ᵤ q)
+protected lemma associated.irreducible [monoid α] {p q : α} (h : p ~ᵤ q)
   (hp : irreducible p) : irreducible q :=
-⟨mt (is_unit_iff_of_associated h).2 hp.1,
+⟨mt h.symm.is_unit hp.1,
   let ⟨u, hu⟩ := h in λ a b hab,
   have hpab : p = a * (b * (u⁻¹ : units α)),
     from calc p = (p * u) * (u ⁻¹ : units α) : by simp
       ... = _ : by rw hu; simp [hab, mul_assoc],
   (hp.is_unit_or_is_unit hpab).elim or.inl (λ ⟨v, hv⟩, or.inr ⟨v * u, by simp [hv]⟩)⟩
 
-lemma irreducible_iff_of_associated [monoid α] {p q : α} (h : p ~ᵤ q) :
+protected lemma associated.irreducible_iff [monoid α] {p q : α} (h : p ~ᵤ q) :
   irreducible p ↔ irreducible q :=
-⟨irreducible_of_associated h, irreducible_of_associated h.symm⟩
+⟨h.irreducible, h.symm.irreducible⟩
 
 lemma associated_mul_left_cancel [comm_cancel_monoid_with_zero α] {a b c d : α}
```

Comment:
```suggestion
lemma associated.mul_left_cancel [comm_cancel_monoid_with_zero α] {a b c d : α}
```
could be worth a shot. Do we like it when one can only dot notate on certain kinds of arguments? If we don't, then we can exchange `h` and `h₁`.

Reply:
I'm not so convinced by this name, since it would suggest something like `associated a b → c * a = c * b → a = b`. How about:
```suggestion
lemma associated.of_mul_left [comm_cancel_monoid_with_zero α] {a b c d : α}
```

Reply:
(And idem for `associated_mul_right_cancel` → `associated.of_mul_right` below)

Reply:
Sure! I think `associated.mul_left_cancel` would be pretty clear (modulo a minimum of intuition about association), but this is fine too.